### DISCUSSION TITLE
Texture Shaders in Cube Geometry

### DISCUSSION THREAD

**Discussion by no-lex** [redeclipse/base#1175] [General]
Texturing in Red Eclipse and its Tesseract engine does more than apply images to surfaces. The game relies on several visual tricks to cheaply simulate real texture depth and albedo (shininess) which are best used when well understood. The game already provides sensible defaults to these textures, but they are not optimal in all circumstances.
#### Diffuse mapping (`stdworld`)
This is the standard color image of the texture, and what the texture browser displays in its tiles. All other shaders also implicity present the diffuse map.
![stdworld](https://user-images.githubusercontent.com/23445929/65181683-6ded9a80-da14-11e9-813b-7b9cfd04f8f9.png)
#### Normal mapping (`bumpworld`)
This is mapping the surface normals (the actual orientation of the surfaces) and calculating how the diffuse map's irradiated light changes because of the orientation of the texture (regions pointing away from you are going to have less area to shine light at you)
![bumpworld](https://user-images.githubusercontent.com/23445929/65181809-bd33cb00-da14-11e9-9207-63ddeb653310.png)
#### Specular highlights (`specworld`)
This creates a specular reflection (where the surface reflects light sources like a mirror) over the entire surface uniformly. The specular reflection borrows none of its color from the underlying texture, and all of it from the light it originally came from.

![specworld](https://user-images.githubusercontent.com/23445929/65181947-0edc5580-da15-11e9-8eff-d3a062aa7ea5.png)
#### Specular mapping (`specmapworld`)
This maps out certain areas of the texture to have more specular reflection than others: some parts of a texture, e.g. metal parts or areas worn smooth, are going to specularly reflect more than other parts of a texture.

A specular map is a single channel grayscale file encoding how reflective an area is for all locations on the surface.

![specmapworld](https://user-images.githubusercontent.com/23445929/65182060-4cd97980-da15-11e9-8263-c92064620e57.png)
#### Parallax mapping (`bumpparallaxworld`)*
* Note that bump is currently required to assign parallax to a texture, even though the two are not necessarily required to be together. There are very few circumstances in which it is possible to justify having parallax and not normal mapping.

Parallax mapping changes how visible parts of the texture are depending on the observers' position relative to them. This is different than normal mapping in that normal mapping merely reduces the intensity of light from regions facing away from you, while parallax mapping reduces its visible size. Together, parallax and normal mapping can create a fairly convincing substitute for actual geometry, though both have visible issues in their approximations at shallow angles.

A parallax map (also called a heightmap, as it encodes vertical position) is a single channel, which can be either its own grayscale file or the alpha channel of the normal map. 

![parallax1](https://user-images.githubusercontent.com/23445929/65182499-45ff3680-da16-11e9-804a-4ac2c3d2d379.png)
![parallax2](https://user-images.githubusercontent.com/23445929/65182498-45ff3680-da16-11e9-8887-e1298bbe4d0b.png)

Note how the size of the dark regions changes depending on the observer's viewpoint.

#### Triplanar mapping (`triplanarworld`)
Triplanar mapping involves mapping the texture from three directions (x,y,z) rather than one and using the information from those three orientations to allow the texture to be mapped accurately at any orientation (rather than having significant error at any orientation other than that of the cube face it occupies). This is most useful for patching seams in compound curvature where no patching of the seam with `voffset` or `vrotate` is possible.

Because triplanar mapping is fairly expensive, it is not recommended to be used unless it is visibly needed.

![triplanarworld](https://user-images.githubusercontent.com/23445929/65183045-71ceec00-da17-11e9-842f-784d6cc146e5.png)

Note that the left cube has a seam running down the middle while the triplanar'd right cube does not.

#### Triplanar detail mapping (`triplanardetailworld`)
While standard triplanar mapping is useful for blending a texture with itself, blending a texture with another according to angle is possibly useful (maybe???) to smoothly transition without blendmaps.

The most plausible use case for `triplanardetailworld` would be to blend three textures together on rough terrain, using the two textures for triplanar detail mapping and laying a blendmap over the top of it for the third. Otherwise, just using blendmap is preferable unless under some very strict map size restrictions or consistency of the blend with respect to angle is critical.

To add the triplanar detail shader to a texture, setting the shader to `triplanardetail*world` and declaring any normal/spec/etc. maps should be done as usual.
The vslot to be blended should then be declared using `texdetail <vslot>` along with other texture commands (e.g. `texscale`) at the end.

![20190920175554](https://user-images.githubusercontent.com/23445929/65365784-3e37c180-dbd1-11e9-9019-7197dfef4873.png)

#### Glow mapping (`glowworld`)
Glow mapping makes certain of the textures always be lit. This is typical for lights fixtures and computer equipment textures as well as other objects which are always lit. 

The glow map is an intensity map of the areas for the engine to glow. It can be given in full RGB color for colored glow effects.

![20190918132721](https://user-images.githubusercontent.com/23445929/65183381-3a147400-da18-11e9-84da-499c1c014339.png)

Note how the light itself in the texture is lit, but the surrounding frame for the light is not, since it is not in the glow map for the texture.

### Using Shaders Ingame
Shader parameters can be changed with the `vshaderparam` command, which has an index in the REEE for easy reference. Note that only one shader can be modified with this command; extensive modifications need to be done in the configuration file.

Note that if you are setting the scale of a shader to 0 with `vshaderparam` you are not disabling it; physically removing the shader from the config file is the only way to get your lost performance back. This means that e.g. `vshaderparam specscale 0 0 0` is not recommended as it leaves performance on the table. 
`vshaderparam` has the following modifications available:
 * envmin : the incidence angle for envmaps
 * envscale : the intensity of envmaps
 * glowcolor :(NOTE: "color" unlike the "colour" used elsewhere) the intensity of glow 
 * parallaxscale : the stops for the parallax effect (recommended to keep values low unless attempting wacky teleporter or liquid effects)
 * pulseglowspeed : frequency of glow pulsing
 * specscale : intensity of specular highlights
 * triplanarbias : intensity of triplanar effect (in x,y,z directions)

### Configuring Shaders
The shaders applied to the currently selected texture are included in the texture stats HUD element, and understanding when and how to change these to better suit your map is important. While the default shaders for most textures are sensible and useful, having the power to change them to help work around specific map issues or provide specific effects is a powerful tool in creating a realistic map.

Maps have their texture list saved in their configuration file, with the textures taking up the entries above the decals, mapmodels, and sounds. They will always be formatted in this way:

```
setshader bumpspecmapparallaxglowworld
texture 0 "acerspyro/light1/light1.png" // 42
texture n "acerspyro/light1/light1_normal.png"
texture s "acerspyro/light1/light1_specular.jpg"
texture g "acerspyro/light1/light1_glow.jpg"
texture z "acerspyro/light1/light1_parallax.png"
texscale 0.5
```
The commented number besides the `texture 0 ...` line is the index in the file (in this case, it's the 43rd texture in the list). It will automatically reset itself once the map is saved, so don't bother messing with it.

The correct order for the `setshader` argument is listed in `config/glsl/world.cfg` where the aliases are defined. However, this format is fairly readable and understandable: this texture here has normals (bumps), specmaps, parallax, and glow enabled on top of the normal diffuse. This means that maps for all of these are required, which are listed below, in the `texture ...` entries.
 * 0 for the diffuse map
 * n for the **n**ormal map
 * s for the **s**pecular map
 * g for the **g**low map
 * z for the height (**z**-coordinate) map

The environment and specular shaders alone do not have maps; however, `specmapenvworld` does require a specular map, as does all other `specmap` shaders.
Additionally, `texscale`, which is just the `vscale` of the texture, is applied at the end to scale the texture properly. This can be changed without issue ingame so it is advised to leave it (and no, changing this won't grant you smaller `vscale` values than .125 nor larger than 8)

### Typical reasons to change default shaders
 * Parallax-defaulting tiles having issues tiling against other textures, as the location of the tile edge changes with orientation
 * Triplanar additions to make corners, organic geometry, or cube voodoo smooth out properly
 * Specular map or envmap additions to add more metalicity to a texture, or conversely,
 * Specular map or envmap removals to remove unwanted unphysical highlights (e.g. an omnidirectional light representing a fixture throwing speculars above it that cannot physically exist)

**Comment by Hirato** [comment]
Just going to leave this here as an additional reference: https://github.com/Hirato/lamiae/wiki/Textures

It's kind of old in various ways, and doesn't cover texcmds i.e. prefixing your texture with `<mad:a/b/c>` or the extended texrotate values
